### PR TITLE
Support for Java 21

### PR DESCRIPTION
**Pull request by LorenzoBettini**
Closes https://github.com/eclipse/xtext/issues/2823

Addresses most of https://github.com/eclipse/xtext/issues/2686

### CODE REVIEW

**Comment by LorenzoBettini** [comment]
A few comments:

- Basically the support of Java 21 from JDT is automatic in Eclipse, taking the latest version of JDT; however, this https://github.com/eclipse/xtext/pull/3037/files#diff-829bb66211b81ee8e7e87794e1319b00452a80ff6916b8feeead5d7e72e6e27eR157 is also required when we use the JDT parser to parse Java files, so that we can correctly detect Java records; that method was introduced in JDT several versions ago, so that's harmless: for sure, it's present in our minimal required version of JDT.
- Concerning Maven the BOM has a profile using the new version of JDT core. I found out that a profile in a BOM CANNOT be activated explicitly with `-P`; in fact, a profile must be present in the aggregator or parent to be explicitly activated. However, good news, a profile in a BOM can still be implicitly activated through a System property! E.g., either at the command line or through the file ".mvn/maven.config" as I do in the IT projects of our Maven plugins, e.g., https://github.com/eclipse/xtext/pull/3037/files#diff-b9a50ce2dde853da7188fbdbf5eac95cf0259c51d108b42360bd1a7dbfda555bR1 For the profile I chose the name `java-21` and the same for the property; if you prefer some other names, please let me know
- I also added a CLI Wizard test for Java 21. I don't know whether it's worthwhile, wdyt?
- As for the rest, it's mostly related to added tests, activated when the runtime is Java 21. I took the chance to refactor a bit standalone incremental builder tests.

One further possibility for Maven projects when selecting Java 21 in the wizard would be to generate the".mvn/maven.config" file. However, I seem to understand that this solution is temporary anyway: when we switch to Java 17 as the minimal requirement, the BOM profile is not needed anymore.

**Comment by LorenzoBettini** [comment]
Another small note: the GitHub Actions `build-maven-artifacts` uses Java 21; while for the full builds there's only an Ubuntu job for Java 21.

**Comment by cdietrich** [comment]
what are the org.eclipse.draw2d  changes for?

**Comment by LorenzoBettini** [comment]
> what are the org.eclipse.draw2d changes for?

It's documented in the tests: it's to access a java record from an existing library. JDK itself doesn't provide any in my findings.

**Comment by cdietrich** [comment]
there are some rare ones e.g.

UnixDomainPrincipal one in MXBean

**Comment by LorenzoBettini** [comment]
> UnixDomainPrincipal

But is that available on all platforms?

I felt safer using something like the one from draw2d, which is in our target platform anyway.

It's a temporary solution until we use Java 21 in Eclipse.

However, I have to remove the lower bound to let it also build with the older platform: in Jenkins, it cannot resolve the latest draw2d

**Comment by LorenzoBettini** [comment]
@cdietrich I removed the lower bound for draw2d to let the build work also on older target platform.

Of course, if you don't like this temporary solution I can also try with UnixDomainPrincipal

**Comment by LorenzoBettini** [comment]
> maybe now that we are in a monorepo. maybe we can reconsider a single testdata bundle for jvm stuff @szarnekow wdyt

I think we already have something like that: `org.eclipse.xtext.xbase.testdata`.
But I couldn't add a Java record there: it wouldn't compile unless we always use Java 21.

That's why I was considering mine as a temporary solution.

**Comment by cdietrich** [comment]
i see. maybe we use draw2d for now and something else later.

**Comment by LorenzoBettini** [comment]
> i see. maybe we use draw2d for now and something else later.

Yes, that's what I was proposing as well.

**Comment by LorenzoBettini** [comment]
Argh! There's a compilation failure in Jenkins:

```
08:40:36  [ERROR] Failed to execute goal org.eclipse.tycho:tycho-compiler-plugin:4.0.7:compile (default-compile) on project org.eclipse.xtext.xbase.ui.tests: Compilation failure: Compilation failure: 
08:40:36  [ERROR] /home/jenkins/agent/workspace/xtext_PR-3037/org.eclipse.xtext.xbase.ui.tests/src/org/eclipse/xtext/xbase/ui/tests/JavaVersionExtendedTest.java:[32] 
08:40:36  [ERROR] 	assertEquals(ClassFileConstants.JDK21, JavaVersion.JAVA21.toJdtClassFileConstant());
08:40:36  [ERROR] 	                                ^^^^^
08:40:36  [ERROR] JDK21 cannot be resolved or is not a field
```

We're using an old version of JDT in that old target platform...

**Comment by cdietrich** [comment]
this is why we in the past partially used reflection

File: org.eclipse.xtend.maven.plugin/Run ITs from Maven.launch [hunk 1]
```diff
@@ -22,6 +22,5 @@
     <booleanAttribute key="org.eclipse.jdt.launching.ATTR_SHOW_CODEDETAILS_IN_EXCEPTION_MESSAGES" value="true"/>
     <booleanAttribute key="org.eclipse.jdt.launching.ATTR_USE_CLASSPATH_ONLY_JAR" value="false"/>
     <booleanAttribute key="org.eclipse.jdt.launching.ATTR_USE_START_ON_FIRST_THREAD" value="true"/>
-    <stringAttribute key="org.eclipse.jdt.launching.JRE_CONTAINER" value="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-1.7/"/>
```
Comment (szarnekow) [inline org.eclipse.xtend.maven.plugin/Run ITs from Maven.launch:25]: Interesting to see what we still had in some of the config files.

**Comment by cdietrich** [comment]
@LorenzoBettini doesnt https://ci.eclipse.org/xtext/job/xtext-monorepo-full-deploy-nightly/ start automatically anymore?

**Comment by LorenzoBettini** [comment]
@cdietrich It is triggered by https://ci.eclipse.org/xtext/job/xtext/job/main/ upon success build.

It has built: https://ci.eclipse.org/xtext/job/xtext-monorepo-full-deploy-nightly/273/

Of course, slowness of JIRO doesn't help ;)

**Comment by cdietrich** [comment]
@LorenzoBettini i started it manually
this is why i am asking

**Comment by cdietrich** [comment]
@LorenzoBettini runniing

JavaSourceLanguageTest with the record i dont see 
org.eclipse.xtext.java.resource.JavaDerivedStateComputer.createType(TypeDeclaration, String)
called

**Comment by LorenzoBettini** [comment]
> @LorenzoBettini i started it manually this is why i am asking

That's strange... indeed I see in the main job 

```
14:21:00  Stage "Trigger Snapshot Deployment" skipped due to when conditional
```

**Comment by LorenzoBettini** [comment]
> > @LorenzoBettini i started it manually this is why i am asking
> 
> That's strange... indeed I see in the main job
> 
> ```
> 14:21:00  Stage "Trigger Snapshot Deployment" skipped due to when conditional
> ```

It worked yesterday though: https://ci.eclipse.org/xtext/job/xtext/job/main/1138/console

**Comment by LorenzoBettini** [comment]
> @LorenzoBettini runniing
> 
> JavaSourceLanguageTest with the record i dont see org.eclipse.xtext.java.resource.JavaDerivedStateComputer.createType(TypeDeclaration, String) called

But the test succeeds and the corresponding class is found.
What do you mean?

**Comment by cdietrich** [comment]
@LorenzoBettini this test wll do nothing.
as org.eclipse.xtext.java.resource.JavaDerivedStateComputer.installStubs(Resource)
will swallow all problems
and not produce a type

**Comment by cdietrich** [comment]
grafik

**Comment by cdietrich** [comment]
=> maybe a installFull vs installStub thing

**Comment by cdietrich** [comment]
gralde trace

```
       at org.eclipse.xtext.java.resource.JavaDerivedStateComputer.createType(JavaDerivedStateComputer.java:130)
        at org.eclipse.xtext.java.resource.JavaDerivedStateComputer.installStubs(JavaDerivedStateComputer.java:107)
        at org.eclipse.xtext.java.resource.JavaResource.lambda$0(JavaResource.java:179)
        at org.eclipse.xtext.java.resource.JavaResource.initializing(JavaResource.java:195)
        at org.eclipse.xtext.java.resource.JavaResource.installStubs(JavaResource.java:178)
        at org.eclipse.xtext.java.resource.JavaResourceDescriptionManager.getResourceDescription(JavaResourceDescriptionManager.java:48)
```

**Comment by cdietrich** [comment]
grafik
no j21 here

**Comment by cdietrich** [comment]
JavaConfig javaConfig = new JavaConfig();
		javaConfig.setJavaSourceLevel(JavaVersion.JAVA21);
		javaConfig.setJavaTargetLevel(JavaVersion.JAVA21);
		javaConfig.attachToEmfObject(rs);

called too late

**Comment by cdietrich** [comment]
see https://github.com/eclipse/xtext/compare/cd-java-source-lang-21?expand=1

**Pull request merged**